### Lab 6 - PRAMOD KUNJUKUNJU SAJI - 8856432

In [142]:
# Importing necessary libraries
import pandas as pd
import numpy as np
from sklearn.linear_model import LogisticRegression
from sklearn.datasets import load_iris
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score, confusion_matrix, precision_score, recall_score, f1_score

1. Using SciKit-Learn, train a logistic regression model on the Iris dataset. Use all four features. Define only 2 labels: `virginica` and `non-virginica`. See the logistic regression notebook presented in class for a demonstration on how to set up these labels

In [143]:
# Loading data from iris dataset
iris_ds = load_iris(as_frame=True)

# Display the keys in the dataset
iris_ds.keys()

dict_keys(['data', 'target', 'frame', 'target_names', 'DESCR', 'feature_names', 'filename', 'data_module'])

In [144]:
# Get first 5 records
iris_ds.data.head()

sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)
0                5.1               3.5                1.4               0.2
1                4.9               3.0                1.4               0.2
2                4.7               3.2                1.3               0.2
3                4.6               3.1                1.5               0.2
4                5.0               3.6                1.4               0.2

In [145]:
print(iris_ds.DESCR)

.. _iris_dataset:

Iris plants dataset
--------------------

**Data Set Characteristics:**

    :Number of Instances: 150 (50 in each of three classes)
    :Number of Attributes: 4 numeric, predictive attributes and the class
    :Attribute Information:
        - sepal length in cm
        - sepal width in cm
        - petal length in cm
        - petal width in cm
        - class:
                - Iris-Setosa
                - Iris-Versicolour
                - Iris-Virginica
                
    :Summary Statistics:

    ============== ==== ==== ======= ===== ====================
                    Min  Max   Mean    SD   Class Correlation
    ============== ==== ==== ======= ===== ====================
    sepal length:   4.3  7.9   5.84   0.83    0.7826
    sepal width:    2.0  4.4   3.05   0.43   -0.4194
    petal length:   1.0  6.9   3.76   1.76    0.9490  (high!)
    petal width:    0.1  2.5   1.20   0.76    0.9565  (high!)
    ============== ==== ==== ======= ===== ===========

In [146]:
# Display the target names
iris_ds.target_names

array(['setosa', 'versicolor', 'virginica'], dtype='<U10')

In [147]:
# Setting the target a Virginica or not. Splitting the data into test and training data
X = iris_ds.data
y = iris_ds.target_names[iris_ds.target] == 'virginica'
X_train, X_test, y_train, y_test = train_test_split(X, y, train_size=0.8, random_state=60)

In [148]:
# Create and fit the logistic regression model
log_reg = LogisticRegression(random_state=60)
log_reg.fit(X_train, y_train)

LogisticRegression(random_state=60)

In [149]:
# Predicting the results
y_pred = log_reg.predict(X_test)

In [150]:
print("Predicted values: ", y_pred)

Predicted values:  [False False False False False False False False False  True False False
  True False  True False  True False False  True  True False False False
 False False False False  True False]


In [151]:
print("Test data: ", y_test)

Test data:  [False False False False False False False False False  True False False
  True False  True False  True False False  True  True False  True False
 False False False False False False]


2. Provide insights about the model prediction. This is an open-ended part. But you can look into questions such as _in which data instances is the model wrong? are there any shared properties for these cases?_ and _how is the model doing, across a set of _performance metrics_ such as _accuracy_ and _confusion metric_.

In [152]:
# Calculating Accuracy score
accuracy = accuracy_score(y_test, y_pred)

# Calculating Precisison score
precision = precision_score(y_test, y_pred)

# Calculating Recall score
recall = recall_score(y_test, y_pred)

# Calculating F1 score
f1_score = f1_score(y_test, y_pred)

# Calculating confusion matrix
cm = confusion_matrix(y_test, y_pred)

print("Accuracy: ", accuracy)
print("Precision: ", precision)
print("Recall: ", recall)
print("F1 Score: ", f1_score)
print("Confusion Matrix: ", cm)

Accuracy:  0.9333333333333333
Precision:  0.8571428571428571
Recall:  0.8571428571428571
F1 Score:  0.8571428571428571
Confusion Matrix:  [[22  1]
 [ 1  6]]


Observations

* Accuracy is the ratio of correctly predicted samples to the total number of samples. In our case, the accuracy is approximately 0.9333, or 93.33%. This indicates that your model predicted correctly for 93.33% of the samples.

* Precision measures the proportion of true positive predictions out of all positive predictions made by the model. In our case, the precision is approximately 0.8571, or 85.71%. This means that out of all the positive predictions made, around 85.71% were correct.

* Recall calculates the proportion of true positive predictions out of all actual positive samples in the dataset. In our case, the recall is approximately 0.8571, or 85.71%. This implies that the model identified around 85.71% of the actual positive samples correctly.

* F1 Score is the harmonic mean of precision and recall. It provides a single metric that balances both precision and recall. In our case, the F1 score is approximately 0.8571, or 85.71%. A high F1 score indicates a good balance between precision and recall.

* The model has achieved reasonably good results based on these metrics, with high accuracy, precision, recall, and F1 score.

Confusion matrix can be interpreted as follows:

* True Positive (TP): There are 22 instances where the actual value is positive and the model correctly predicted them as positive.

* True Negative (TN): There are 6 instances where the actual value is negative and the model correctly predicted them as negative.

* False Positive (FP): There is 1 instance where the actual value is negative, but the model predicted it as positive.

* False Negative (FN): There is 1 instance where the actual value is positive, but the model predicted it as negative.

* The model has a high number of true positives and true negatives, indicating good performance overall. However, the presence of a few false positives and false negatives suggests some errors in the predictions for those particular cases.